[OM-88273]: Initial version of GitopsConfig CRD

amd-ibm opened the pull request:
## Intent

Introduce new CRD to allow override of default Gitops configuration.

## Background

Converted project to multi-group project to enable creation of new gitops.turbonomic.io group.
Moved existing policy-related CRDs into new multi-group package structure.
Added initial version of `GitopsConfig` CRD.

## Testing

Deployed CRD and tested enums, optionals, etc with sample CR.

amd-ibm committed:
[OM-88273]: Converted project to multi-group project to enable creation of new gitops.turbonomic.io group.
Moved existing policy-related CRDs into new multi-group package structure.
Added initial version of GitOpsMode CRD.

pankajbatra1 commented on config/crd/bases/gitops.turbonomic.io.turbonomic.io_gitopsmodes.yaml:
```diff
@@ -0,0 +1,107 @@
+                      description: Specifies the applications that the commit mode
+                        should apply to
+                      items:
+                        type: string
```
What do these strings correspond to?  Are there other resources that should have the same name?

amd-ibm replied:
This is the array of applications that this configuration should apply to. The property name in this case is "applications", which is an array of strings. These are essentially validation rules.

ading1977 commented on apis/gitops.turbonomic.io/v1alpha1/gitopsmode_types.go:
```diff
@@ -0,0 +1,98 @@
+//+kubebuilder:subresource:status
+
+// GitOpsMode is the Schema for the gitopsmodes API
+type GitOpsMode struct {
```
I wonder if it makes more sense to just name this `GitOps`, or `GitCommit`? Because the configuration contains more than the commit mode, it also contains the credential.

amd-ibm replied:
I'm open to any name, really. I couldn't come up with a good one. This is what Irfan suggested.

ading1977 replied:
I see, @irfanurrehman , what is your thoughts?

amd-ibm replied:
Since I'm going to move this to a more generic `devops` group, I think I'll rename this as well to the more generic `GitOps`.

irfanurrehman replied:
Apologies for not pitching in earlier. I think this is best described as `GitopsConfig`.

irfanurrehman replied:
Also please mark that `Gitops` is widely used as one word so only `G` is capitalised.

amd-ibm replied:
Renamed to `GitopsConfig`

ading1977 commented on apis/gitops.turbonomic.io/v1alpha1/gitopsmode_types.go:
```diff
@@ -0,0 +1,98 @@
+	// Valid values are:
+	// - "direct": actions will produce commit directly within the underlying repository without creating a pull/merge request;
+	// - "pr": actions will result in a pull/merge request being creating within the underlying repository
+	CommitMode CommitMode `json:"commitMode"`
```
Does it make sense to make this optional, and provide a default value of `pr`  like we discussed earlier?

amd-ibm replied:
My understanding is that the default configuration will still come from the parameters passed to kubeturbo and that these would _only_ be used to override that configuration. That's why I assumed this should be required. I'm not opposed to adding the default value, however, if you think that's preferred.

amd-ibm replied:
Updated `commitMode` to be optional with default value `pr`.

ading1977 commented on apis/gitops.turbonomic.io/v1alpha1/gitopsmode_types.go:
```diff
@@ -0,0 +1,98 @@
+	// +optional
+	Credentials GitOpsCredentials `json:"credentials,omitempty"`
+	// Specifies the applications that the commit mode should apply to
+	Applications []string `json:"applications"`
```
Shall we add a validation here to make sure the list of applications is **NOT** empty?

amd-ibm replied:
Sure. I'll get that added.

amd-ibm replied:
Added validations to ensure list of apps is nonempty.

ading1977 commented:
Thanks for converting `turbo-crd` from single group to multi-group!

Will this change affect the existing logic in `kubeturbo` that parses the turbo policy? There is nothing wrong with this, we just need to update the code. Just want to confirm.

amd-ibm commented:
> Thanks for converting `turbo-crd` from single group to multi-group!
> 
> Will this change affect the existing logic in `kubeturbo` that parses the turbo policy? There is nothing wrong with this, we just need to update the code. Just want to confirm.

Looks like I'll need to update all of the imports to use the new pathing.

kuanf commented on apis/gitops.turbonomic.io/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "gitops.turbonomic.io.turbonomic.io", Version: "v1alpha1"}
```
"gitops.turbonomic.io.turbonomic.io". I don't think we want this as group.
Actually, do we want to limit this group to gitops? or shall we change it to devops to accommodate other CRs?

amd-ibm replied:
I hadn't noticed that. I'll look into why that was duplicated.

amd-ibm replied:
Looks like that was the result of passing `--group gitops.turbonomic.io` to the `kubebuilder create...` command. Should have just passed `--group gitops`.

amd-ibm replied:
As far as moving to a `devops` group, I think that makes sense. I'll update it.

amd-ibm replied:
Updated to be within the `devops` group.

kuanf commented:
To highlight, let's fix the group and regenerate manifests!

amd-ibm committed:
Renamed CR to GitOps.
Moved to more generic "devops" group.
Added additional validation.

amd-ibm committed:
Fixed policy package.

amd-ibm committed:
Fixed policy path in project file.

amd-ibm committed:
Fixed commitMode requirement.

ading1977 commented:
Do you want to run this change with Eva (maybe in a sidebar during standup)? But otherwise, I am good with the changes.

irfanurrehman commented on apis/devops/v1alpha1/gitops_types.go:
```diff
@@ -0,0 +1,100 @@
+// GitOpsSpec defines the desired state of GitOps
+type GitOpsSpec struct {
+	// Overrides the default GitOps configuration with custom configuration for the specified app(s).
+	Configutaion []GitOpsConfiguration `json:"configuration"`
```
I would recommend to use the short name for json value atleast
s/`json:"configuration"`/`json:"config"`

You can choose to change the structure names also.
So if you take my suggestion to change the name of CR to `GitopsConfig`
`GitOpsSpec` will be come `GitOpsConfigSpec`
and other changes would be

```
type GitOpsConfigSpec struct {
	// Overrides the default GitOps configuration with custom configuration for the specified app(s).
	Config []Configuration `json:"config"`
}


type Configuration struct {
	// Specifies the GitOps commit mode.
	// Valid values are:
....
....


```

amd-ibm replied:
Shortened the config and apps property names.

irfanurrehman commented on apis/devops/v1alpha1/gitops_types.go:
```diff
@@ -0,0 +1,100 @@
+	Applications []string `json:"applications"`
+}
+
+type GitOpsCredentials struct {
```
You dont need to have GitOps appended to every structure.. :-)
This is simply `Credentials`

amd-ibm replied:
Dropped the `Gitops` prefixes from the structs.

irfanurrehman reviewed (commented):
Thanks for the update @amd-ibm.
Have few naming suggestions.

irfanurrehman commented:
Pasting [this](https://github.com/turbonomic/turbo-crd/pull/4#discussion_r1013988946) as general comment, so that you don't miss it above

amd-ibm committed:
Renamed custom resource to "GitopsConfig"
Shortened "config" and "apps" property names.
Removed "GitOps" prefix from types.

irfanurrehman reviewed (approved):
Thanks @amd-ibm. I am not sure, but if there are commits which need squashing, please do that before the merge.

amd-ibm commented:
Will re-open in new [turbo-gitops](https://github.com/turbonomic/turbo-gitops) repo.
